File: frontend/src/stores/aiChat.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import axios from 'axios'

import { aiChatService } from '@/services/aiChatService'
import { useAuthStore } from '@/stores/auth'
import { DEMO_AI_DISCLAIMER, buildDemoAiReply, demoAiConversations } from '@/stores/demoData'
import type { AIConversation, AIConversationDetail, AIUsage } from '@/types/aiChat'

function cloneDemoConversations(): AIConversationDetail[] {
  return demoAiConversations.map((conversation) => ({
    ...conversation,
    messages: conversation.messages.map((message) => ({ ...message })),
  }))
}

function countDemoUsage(conversations: AIConversationDetail[]): number {
  return conversations.reduce(
    (total, conversation) => total + conversation.messages.filter((m) => m.role === 'user').length,
    0,
  )
}

export const useAiChatStore = defineStore('aiChat', () => {
  const demoConversations = ref<AIConversationDetail[]>(cloneDemoConversations())

  const conversations = ref<AIConversation[]>([])
  const activeConversation = ref<AIConversationDetail | null>(null)
  const usage = ref<AIUsage | null>(null)
  const loading = ref(false)
  const sending = ref(false)
  const error = ref('')

  async function loadConversations() {
    const auth = useAuthStore()
    loading.value = true
    error.value = ''
    try {
      conversations.value = auth.isDemoMode
        ? demoConversations.value.map(({ id, title, created_at }) => ({ id, title, created_at }))
        : await aiChatService.listConversations()
    } catch {
      error.value = 'Não foi possível carregar suas conversas.'
    } finally {
      loading.value = false
    }
  }

  async function loadUsage() {
    const auth = useAuthStore()
    if (auth.isDemoMode) {
      usage.value = { plan: 'premium', used: countDemoUsage(demoConversations.value), limit: 1000 }
      return
    }
    try {
      usage.value = await aiChatService.getUsage()
    } catch {
      usage.value = null
    }
  }

  async function selectConversation(id: string) {
    const auth = useAuthStore()
    loading.value = true
    error.value = ''
    try {
      if (auth.isDemoMode) {
        activeConversation.value = demoConversations.value.find((c) => c.id === id) ?? null
        return
      }
      activeConversation.value = await aiChatService.getConversation(id)
    } catch {
      error.value = 'Não foi possível carregar a conversa.'
    } finally {
      loading.value = false
    }
  }

  async function createConversation() {
    const auth = useAuthStore()
    if (auth.isDemoMode) {
      const conversation: AIConversationDetail = {
        id: `demo-conversation-${Date.now()}`,
        title: 'Nova conversa',
        created_at: new Date().toISOString(),
        messages: [],
      }
      demoConversations.value.unshift(conversation)
      await loadConversations()
      activeConversation.value = conversation
      return conversation
    }
    const conversation = await aiChatService.createConversation()
    await loadConversations()
    activeConversation.value = { ...conversation, messages: [] }
    return conversation
  }

  async function deleteConversation(id: string) {
    const auth = useAuthStore()
    if (auth.isDemoMode) {
      demoConversations.value = demoConversations.value.filter((c) => c.id !== id)
    } else {
      await aiChatService.deleteConversation(id)
    }
    if (activeConversation.value?.id === id) {
      activeConversation.value = null
    }
    await loadConversations()
  }

  async function sendMessage(content: string) {
    const auth = useAuthStore()
    if (!activeConversation.value) return

    sending.value = true
    error.value = ''
    try {
      if (auth.isDemoMode) {
        const conversation = activeConversation.value
        conversation.messages.push({
          id: `demo-msg-${Date.now()}`,
          role: 'user',
          content,
          created_at: new Date().toISOString(),
          disclaimer: null,
        })
        if (conversation.title === 'Nova conversa') {
          conversation.title = content.slice(0, 80)
        }
        await new Promise((resolve) => setTimeout(resolve, 500))
        conversation.messages.push({
          id: `demo-msg-${Date.now() + 1}`,
          role: 'assistant',
          content: buildDemoAiReply(content),
          created_at: new Date().toISOString(),
          disclaimer: DEMO_AI_DISCLAIMER,
        })
        await loadUsage()
        return
      }

      activeConversation.value.messages.push({
        id: `pending-${Date.now()}`,
        role: 'user',
        content,
        created_at: new Date().toISOString(),
        disclaimer: null,
      })
      const reply = await aiChatService.sendMessage(activeConversation.value.id, content)
      activeConversation.value.messages.push(reply)
      await loadUsage()
    } catch (err) {
      activeConversation.value.messages.pop()
      error.value =
        axios.isAxiosError(err) && err.response?.status === 429
          ? 'Você atingiu o limite de mensagens de IA do seu plano neste mês.'
          : 'Não foi possível enviar sua mensagem. Tente novamente.'
    } finally {
      sending.value = false
    }
  }

  return {
    conversations,
    activeConversation,
    usage,
    loading,
    sending,
    error,
    loadConversations,
    loadUsage,
    selectConversation,
    createConversation,
    deleteConversation,
    sendMessage,
  }
})

File: backend/app/api/v1/ai_chat.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.v1.deps import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.ai_chat import (
    ConversationDetailRead,
    ConversationRead,
    MessageCreate,
    MessageRead,
    UsageRead,
)
from app.services import ai_chat_service

router = APIRouter(prefix="/ai/chat", tags=["ai-chat"])


@router.get("/usage", response_model=UsageRead)
async def get_usage(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    used, limit = await ai_chat_service.get_usage(db, current_user)
    plan = current_user.subscription.plan if current_user.subscription else "free"
    return UsageRead(plan=plan, used=used, limit=limit)


@router.get("/conversations", response_model=list[ConversationRead])
async def list_conversations(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await ai_chat_service.list_conversations(db, current_user.id)


@router.post("/conversations", response_model=ConversationRead, status_code=status.HTTP_201_CREATED)
async def create_conversation(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await ai_chat_service.create_conversation(db, current_user.id)


@router.get("/conversations/{conversation_id}", response_model=ConversationDetailRead)
async def get_conversation(
    conversation_id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    from app.schemas.ai_chat import DISCLAIMER

    try:
        conversation = await ai_chat_service.get_conversation(db, current_user.id, conversation_id)
    except ai_chat_service.ConversationNotFoundError:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Conversa não encontrada")

    return ConversationDetailRead(
        id=conversation.id,
        title=conversation.title,
        created_at=conversation.created_at,
        messages=[
            MessageRead(
                id=msg.id,
                role=msg.role,
                content=msg.content,
                created_at=msg.created_at,
                disclaimer=DISCLAIMER if msg.role == "assistant" else None,
            )
            for msg in conversation.messages
        ],
    )


@router.delete("/conversations/{conversation_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_conversation(
    conversation_id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    try:
        await ai_chat_service.delete_conversation(db, current_user.id, conversation_id)
    except ai_chat_service.ConversationNotFoundError:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Conversa não encontrada")


@router.post("/conversations/{conversation_id}/messages", response_model=MessageRead)
async def send_message(
    conversation_id: UUID,
    payload: MessageCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    try:
        message, disclaimer = await ai_chat_service.send_message(
            db, current_user, conversation_id, payload.content
        )
    except ai_chat_service.ConversationNotFoundError:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Conversa não encontrada")
    except ai_chat_service.AIRateLimitExceededError:
        raise HTTPException(
            status_code=status.HTTP_429_TOO_MANY_REQUESTS,
            detail="Limite de mensagens de IA do seu plano atingido neste mês",
        )

    return MessageRead(
        id=message.id,
        role=message.role,
        content=message.content,
        created_at=message.created_at,
        disclaimer=disclaimer,
    )
